### ISSUE TITLE
Files that are claimed to contain errors (but don't) are still changed

### ISSUE THREAD

**Issue by sebastianbergmann** [FriendsOfPHP/PHP-CS-Fixer#3205]
I am using the latest stable release:
```
$ php-cs-fixer --version
PHP CS Fixer 2.8.0 Round Numbers by Fabien Potencier and Dariusz Ruminski (89e7b08)
```
When I invoke `php-cs-fixer` for PHPUnit (https://github.com/sebastianbergmann/phpunit/) then I get the following:
```
$ php-cs-fixer fix
Loaded config default from "/usr/local/src/phpunit/.php_cs".

Fixed all files in 4.629 seconds, 22.000 MB memory used

Files that were not fixed due to errors reported during fixing:
   1) /usr/local/src/phpunit/src/Util/TextTestListRenderer.php
   2) /usr/local/src/phpunit/src/Util/XmlTestListRenderer.php
```
I have no idea what errors `php-cs-fixer` thinks these files have. `php -l` thinks they're fine.
```
$ git status
On branch master
Your branch is up-to-date with 'origin/master'.

nothing to commit, working tree clean
```
As you can see above, `php-cs-fixer` did not change any files.
```
$ php-cs-fixer fix
Loaded config default from "/usr/local/src/phpunit/.php_cs".
Using cache file ".php_cs.cache".
   1) src/Util/TextTestListRenderer.php
   2) src/Util/XmlTestListRenderer.php

Fixed all files in 0.026 seconds, 8.000 MB memory used
```
However, running `php-cs-fixer` again changes the files that were reported to contain errors (see above) before:
```
$ git status
On branch master
Your branch is up-to-date with 'origin/master'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   src/Util/TextTestListRenderer.php
	modified:   src/Util/XmlTestListRenderer.php

no changes added to commit (use "git add" and/or "git commit -a")
```

```
$ git diff
diff --git a/src/Util/TextTestListRenderer.php b/src/Util/TextTestListRenderer.php
index 4ee6ac769..be1d2c3f7 100644
--- a/src/Util/TextTestListRenderer.php
+++ b/src/Util/TextTestListRenderer.php
@@ -20,7 +20,8 @@ public function render(TestSuite $suite): string
     {
         $buffer = 'Available test(s):' . PHP_EOL;
 
-        foreach (new \RecursiveIteratorIterator($suite->getIterator()) as $test) {
+        foreach {
+            (new \RecursiveIteratorIterator($suite->getIterator()) as $test) {
             if ($test instanceof TestCase) {
                 $name = \sprintf(
                     '%s::%s',
@@ -39,6 +40,7 @@ public function render(TestSuite $suite): string
             );
         }
 
-        return $buffer;
+            return $buffer;
+        }
     }
 }
diff --git a/src/Util/XmlTestListRenderer.php b/src/Util/XmlTestListRenderer.php
index 2d758d529..f7a8c2c66 100644
--- a/src/Util/XmlTestListRenderer.php
+++ b/src/Util/XmlTestListRenderer.php
@@ -27,7 +27,8 @@ public function render(TestSuite $suite): string
 
         $currentTestCase = null;
 
-        foreach (new \RecursiveIteratorIterator($suite->getIterator()) as $test) {
+        foreach {
+            (new \RecursiveIteratorIterator($suite->getIterator()) as $test) {
             if ($test instanceof TestCase) {
                 if (\get_class($test) !== $currentTestCase) {
                     if ($currentTestCase !== null) {
@@ -71,12 +72,13 @@ public function render(TestSuite $suite): string
             }
         }
 
-        if ($currentTestCase !== null) {
+            if ($currentTestCase !== null) {
+                $writer->endElement();
+            }
+
             $writer->endElement();
         }
 
-        $writer->endElement();
-
         return $writer->outputMemory();
     }
 }
```

```
$ php -l src/Util/TextTestListRenderer.php
PHP Parse error:  syntax error, unexpected '{', expecting '(' in src/Util/TextTestListRenderer.php on line 23
Errors parsing src/Util/TextTestListRenderer.php
```

```
$ php -l src/Util/XmlTestListRenderer.php
PHP Parse error:  syntax error, unexpected '{', expecting '(' in src/Util/XmlTestListRenderer.php on line 30
Errors parsing src/Util/XmlTestListRenderer.php
```
Now these files contain errors that look similar to #3202.

**Comment by SpacePossum** [comment]
The fixer tries to change PHP source code without linting errors into code that has linting error (3202 does indeed prove that).
The error message states;
>Files that were not fixed due to errors reported during fixing

Which means that; after fixing the fixer found out the result was not valid PHP and therefor does not apply the changes (i.e. not writes back the change to file).

I'll leave 3202 to @julienfalque , however I'm puzzled about why the fixer _did_ apply the second time it was run...

**Comment by julienfalque** [comment]
This should be fixed by #3208.

> after fixing the fixer found out the result was not valid PHP and therefor does not apply the changes (i.e. not writes back the change to file).

@SpacePossum I tried running the fixer on a project and the same bug occured, but the changes were applied anyway, so no linting check before it seems. Maybe we should keep this issue open despite it's a duplicate of #3202 until we figured out what happens?

**Comment by SpacePossum** [comment]
> Maybe we should keep this issue open [...]until we figured out what happens?

:+1: for keeping this one open to check why the fixer write back code it knows doesn't lint.

Thanks for all the other work on the ticket @julienfalque :)

**Comment by keradus** [comment]
This reveals to use 2 bugs actually. Fixed in #3208 and #3217 , which just got merged (not yet released).
Thanks @sebastianbergmann
